### PR TITLE
resolves #24: fix async binary transfer, fix sync content length

### LINKED ISSUE #24: Fix async binary transfer
Async only transfers text/utf8.  Lets fix this!

Also, your content-length test is wrong because of a misplaced \n trim.

### PR DESCRIPTION
**Pull request by djencks**
Fixes async binary transfer, runs transfer tests both async and sync, and fixes a post content trimming mistake.

### CODE REVIEW

File: lib/XMLHttpRequest.js [hunk 1]
```diff
@@ -117,6 +119,8 @@ exports.XMLHttpRequest = function () {
   this.withCredentials = false
   // "text", "arraybuffer", "blob", or "document", depending on your data needs.
   // Note, setting xhr.responseType = '' (or omitting) will default the response to "text".
+  // Actual accepted values for async are "text", "arraybuffer", or "buffer".
+  // Other values will result in "buffer" for async and "arraybuffer" (Uint8Array) for sync
```
Comment (Mogztter) [inline lib/XMLHttpRequest.js:123]: > The XMLHttpRequest property responseType is an enumerated string value specifying the type of data contained in the response. It also lets the author change the response type. If an empty string is set as the value of responseType, the default value of text is used.

Are you sure that invalid values are treated as `buffer` for async and `arraybuffer` for sync. I need to check but I would expect the default value to be used `text`.

File: tests/test-sync-request.js [hunk 2]
```diff
@@ -7,112 +7,113 @@ chai.use(dirtyChai)
 const expect = chai.expect
 const XMLHttpRequest = require('../lib/XMLHttpRequest').XMLHttpRequest
 
-describe('XMLHttpRequest sync request', () => {
```
Comment (Mogztter) [inline tests/test-sync-request.js:10]: The file is named `test-sync-request.js`. If we want to test both sync and async requests we should rename it. Alternatively we could create a dedicated file for async requests (and share some code between the two tests suite).

File: package.json [hunk 3]
```diff
@@ -7,11 +7,13 @@
   ],
   "scripts": {
     "test": "mocha tests/test-*.js",
-    "lint": "standard tests/**.js lib/**.js example/**.js"
+    "lint": "standard tests/**.js lib/**.js example/**.js",
+    "lint-fix": "standard tests/**.js lib/**.js example/**.js --fix"
```
Comment (Mogztter) [inline package.json:11]: Do not repeat the entire command, you can reference it:

```suggestion
    "lint-fix": "npm run lint -- --fix"
```

File: lib/XMLHttpRequest.js (hunk 1 above)
Comment (djencks) [reply lib/XMLHttpRequest.js:123]: All values except 'text' and '' result in Buffer (which is an Uint8ArrayBuffer). (after a slight change).  Wording to be changed....

File: tests/test-sync-request.js (hunk 2 above)
Comment (djencks) [reply tests/test-sync-request.js:10]: yes. I'm also trying to enable https testing.

File: package.json (hunk 3 above)
Comment (djencks) [reply package.json:11]: yes.

**Comment by djencks** [comment]
I added https tests and the ability to set rejectUnauthorized, with a test.... demonstrating a problem with sync errors.

**Comment by Mogztter** [comment]
This looks great but could you please create one pull request per change otherwise it's hard to review:

- `JSON.stringify(error)`
- `rejectUnauthorized`
- Fix `content-length` with trailing `\n`
- Fix async binary transfer

Thanks

**Comment by djencks** [comment]
Separate pull requests would be a lot of work and don't reflect the history, which was that I made binary transfer work in a project, wrote some tests, and found some fairly egregious bugs which I fixed.  Perhaps I could annotate the changes instead?

**Comment by Mogztter** [comment]
> Separate pull requests would be a lot of work and don't reflect the history,

The problem is that there's no history, all changes are already squashed into a single commit.

> Perhaps I could annotate the changes instead?

Yes that could help.
But it feels like you are fixing multiple things at once. So I would still like to have one commit/one pull request per fix/change.

File: lib/request.js [hunk 4]
```diff
@@ -66,6 +66,8 @@ function doRequest (options, data) {
       // Even though executed by name, the first argument is still "node",
       // the second the script name. The third is the string we want.
       data = Buffer.from(process.argv[2] || '', encoding)
+      // There will be a trailing \n from the user hitting enter. Get rid of it.
+      data = data.replace(/\n$/, '')
```
Comment (djencks) [inline lib/request.js:70]: This code is obviously for TTY input, and it was located for non-TTY input.  As a result of being applied in the wrong place, the actual length of the test string, 30, was reduced in the test to 29.

File: lib/request.js [hunk 5]
```diff
@@ -81,10 +83,12 @@ function doRequest (options, data) {
         }
       })
       process.stdin.on('end', async function () {
-        // There will be a trailing \n from the user hitting enter. Get rid of it.
-        data = data.replace(/\n$/, '')
```
Comment (djencks) [inline lib/request.js:85]: Here's the code in the wrong place.

File: tests/test-request.js [hunk 6]
```diff
@@ -0,0 +1,162 @@
...
+                      responseText = xhr.responseText
+                    }
+                    resolve()
+                  }
+                }
+                xhr.send(' \n{  "message":  "hello" }  \r\n')
+              }
+            })
+          })
+          expect(responseText).to.equal('30')
```
Comment (djencks) [inline tests/test-request.js:118]: Here's the correct length of the test message.

**Review by djencks — COMMENTED** [review]
Correct content-length test.

File: lib/request.js [hunk 7]
```diff
@@ -81,10 +83,12 @@ function doRequest (options, data) {
...
-        const result = await doRequest(options, data)
-        console.log(JSON.stringify(result))
+        try {
+          const result = await doRequest(options, data)
+          console.log(JSON.stringify(result))
+        } catch (e) {
+          console.log(JSON.stringify({ error: e }))
+        }
```
Comment (djencks) [inline lib/request.js:91]: This adds some pretty obvious error handling, the need for which was made obvious by the tests. The following catch is in a different function, so this error got swallowed silently.

**Review by djencks — COMMENTED** [review]
Fix error handling in sync case.

File: lib/XMLHttpRequest.js [hunk 8]
```diff
@@ -431,29 +436,58 @@ exports.XMLHttpRequest = function () {
...
+          response.on('end', function () {
+            // buffers are Uint8Array instances
+            self.response = Buffer.concat(self.responseBinary)
+            if (sendFlag) {
+              // Discard the end event if the connection has been aborted
+              setState(self.DONE)
+              sendFlag = false
+            }
+          })
+        }
```
Comment (djencks) [inline lib/XMLHttpRequest.js:491]: This is the functional change to handle binary transfers async.

File: lib/XMLHttpRequest.js [hunk 9]
```diff
@@ -490,7 +524,7 @@ exports.XMLHttpRequest = function () {
         response = result.data
         self.status = result.data.statusCode
         if (encoding === 'binary') {
-          self.response = Uint8Array.from(result.data.binary.data).buffer
```
Comment (djencks) [inline lib/XMLHttpRequest.js:493]: This conversion was unnecessary as a Buffer is a Uint8Array.

File: lib/XMLHttpRequest.js [hunk 10]
```diff
@@ -504,7 +538,7 @@ exports.XMLHttpRequest = function () {
    */
   this.handleError = function (error) {
     this.status = 0
-    this.statusText = error
```
Comment (djencks) [inline lib/XMLHttpRequest.js:507]: (this belongs in the previous review) An error is not a string, so stringify it.  For some reason the sync and async errors have slightly different info, but for the cert test they both have enough in common to recognize.

**Review by djencks — COMMENTED** [review]
This is the functional change.

File: lib/XMLHttpRequest.js [hunk 11]
```diff
@@ -170,7 +173,8 @@ exports.XMLHttpRequest = function () {
       url: url.toString(),
       async: (typeof async !== 'boolean' ? true : async),
       user: user || null,
-      password: password || null
+      password: password || null,
+      rejectUnauthorized: (typeof rejectUnauthorized !== 'boolean' ? true : rejectUnauthorized)
```
Comment (djencks) [inline lib/XMLHttpRequest.js:177]: introduce rejectUnauthorized setting with default of true to enable easy https testing without a valid certificate.

File: lib/XMLHttpRequest.js [hunk 12]
```diff
@@ -382,7 +386,8 @@ exports.XMLHttpRequest = function () {
       method: settings.method,
       headers: headers,
       agent: false,
-      withCredentials: self.withCredentials
+      withCredentials: self.withCredentials,
+      rejectUnauthorized: settings.rejectUnauthorized
```
Comment (djencks) [inline lib/XMLHttpRequest.js:390]: Transfer the rejectUnauthorized setting

File: tests/test-request.js [hunk 13]
```diff
@@ -0,0 +1,162 @@
...
+                responseText = xhr.statusText
+                resolve()
+              }
+              await xhr.send('ping! "from client"')
+            }
+          })
+        })
+        expect(responseText).to.equal('{"code":"CERT_HAS_EXPIRED"}')
+      } finally {
+        child.kill()
```
Comment (djencks) [inline tests/test-request.js:158]: Check that with rejectUnauthorized true (default) we can't communicate over https with a server with an expired certificate.

**Review by djencks — COMMENTED** [review]
rejectUnauthorized configuration.

File: lib/XMLHttpRequest.js (hunk 11 above)
Comment (Mogztter) [reply lib/XMLHttpRequest.js:177]: It's not part of [`XMLHttpRequest.open`](https://developer.mozilla.org/en-US/docs/Web/API/XMLHttpRequest/open). I'm not sure if it's possible to do an insecure request with `XMLHttpRequest`...

Comment (djencks) [reply lib/XMLHttpRequest.js:177]: Hmmm.... this class already diverges in capabilities from XMLHttpRequest, and this is an optional last parameter defaulting to secure.
Do you have an alternate way to test https? If so I'm happy to try it out.

Comment (Mogztter) [reply lib/XMLHttpRequest.js:177]: > Hmmm.... this class already diverges in capabilities from XMLHttpRequest, 

Could you expand on that? The original goal of this library was to provide a compatible implementation of `XMLHttpRequest` in Node so we should try to stick to it.

> Do you have an alternate way to test https? If so I'm happy to try it out.

I guess we will need to somehow override/mock the settings in the test without altering the API (production code).
Maybe Sinon.js could help.

File: lib/XMLHttpRequest.js (hunk 9 above)
Comment (Mogztter) [reply lib/XMLHttpRequest.js:493]: Reading the code again, I don't even remember how `data` is defined. I can see that we assigned the Buffer to `result.data.binary` but I don't explain the last `.data`.

Anyway I think you're right and the code can be simplified.

Comment (Mogztter) [reply lib/XMLHttpRequest.js:493]: `result.data.binary` is stringify so `result.data.binary.data` is an `Array` not a `Buffer` or an `Uint8Array`.
The specification says that it should return a `ArrayBuffer`.

File: lib/XMLHttpRequest.js (hunk 11 above)
Comment (djencks) [reply lib/XMLHttpRequest.js:177]: It doesn't support most of the return types. What's the problem with supporting a non-standard extension that doesn't interfere with anything if you don't use it?
I didn't look hard, but it does't look to me as if sinon.js actually uses https requests.
If you insist on removing the open parameter, I'll see if there's some way to directly set something on the instance to accomplish the same thing.

File: lib/XMLHttpRequest.js [hunk 14]
```diff
@@ -504,7 +538,7 @@ exports.XMLHttpRequest = function () {
    */
   this.handleError = function (error) {
     this.status = 0
-    this.statusText = error
+    this.statusText = JSON.stringify(error)
```
Comment (Mogztter) [inline lib/XMLHttpRequest.js:541]: This is less wrong but still wrong, see #30 
Could you please revert this change? I've addressed this issue in #31.
Ideally, we should infer the status text from the error but that would require a bit of work. The reason is that we would have to generate the exhaustive list of all the errors that might happen :disappointed:

File: lib/XMLHttpRequest.js (hunk 11 above)
Comment (Mogztter) [reply lib/XMLHttpRequest.js:177]: > I didn't look hard, but it does't look to me as if sinon.js actually uses https requests.

Sinon.js can create stubs and mocks, so you could declare a `getHttpOptions` function and stub it in the test to pass the `rejectUnauthorized`.

> It doesn't support most of the return types. 

If some use cases are not supported I think it's fair but we should try to adhere to the specification when we implement something.

> What's the problem with supporting a non-standard extension that doesn't interfere with anything if you don't use it?

It does interfere since the function now has 6 parameters. Arguably, it's not a big deal but I don't feel like it's the right way to do.
Doing an insecure request is simply not available/possible in `XMLHttpRequest`.

As you said, we only need it in the tests suite (to do an HTTPS request without a valid certificate) so we should not alter the production code.

File: lib/request.js (hunk 4 above)
Comment (Mogztter) [reply lib/request.js:70]: I've extracted this change in #29 
Could you please revert it in this pull request?

File: lib/request.js (hunk 7 above)
Comment (Mogztter) [reply lib/request.js:91]: > The following catch is in a different function, so this error got swallowed silently.

Indeed, nice catch :+1: 
Could you please annotate the corresponding test? I will extract this change in a dedicated pull request.

File: lib/XMLHttpRequest.js (hunk 11 above)
Comment (djencks) [reply lib/XMLHttpRequest.js:177]: Apparently we can avoid the parameter by setting
```
process.env["NODE_TLS_REJECT_UNAUTHORIZED"] = 0;
```
in the appropriate tests.

File: tests/test-request.js (hunk 6 above)
Comment (djencks) [reply tests/test-request.js:118]: I've rebased on top of #29.

File: lib/XMLHttpRequest.js (hunk 9 above)
Comment (djencks) [reply lib/XMLHttpRequest.js:493]: Fixed (after a lot of learning) in a new commit.

File: lib/XMLHttpRequest.js (hunk 14 above)
Comment (djencks) [reply lib/XMLHttpRequest.js:541]: I think #31 isn't quite right either, I've addressed this in a new commit.

Comment (djencks) [reply lib/XMLHttpRequest.js]: BTW, this was previously untested, and is now tested in the last `test-request` test in 
```
        expect(responseText).to.equal('')
        expect(errorText).to.equal('{"code":"CERT_HAS_EXPIRED"}')
```

File: lib/XMLHttpRequest.js (hunk 9 above)
Comment (Mogztter) [reply lib/XMLHttpRequest.js:491]: > Fixed (after a lot of learning) in a new commit.

:+1: 
I think we should add a comment to preserve this knowledge. We should explain why we are using `new Uint8Array()` instead `Uint8Array.from` (or if we could use one or the other).
And we should also state the type of `result.data.binary`: "result.data.binary is stringify so result.data.binary.data is an Array"

**Comment by Mogztter** [comment]
All your changes has been integrated in other pull requests. Thanks for your fine work :ok_hand: 
Could you please test against master to make sure that everything is working as expected?

Please note that I did a few changes to comply with the `XMLHttpRequest`.

**Comment by Mogztter** [comment]
Closing, thanks again for all these fixes! :clap:
